[ICE]: `generics_of: unexpected node kind Crate(Mod..`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:212:14: generics_of: unexpected node kind Crate(Mod { spans: ModSpans { inner_span: /tmp/im/a.rs:1:1: 20:57 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ a[ad07]::std) }, ItemId { owner_id: DefId(0:2 ~ a[ad07]::{use#0}) }, ItemId { owner_id: DefId(0:3 ~ a[ad07]::ActuallySuper) }, ItemId { owner_id: DefId(0:5 ~ a[ad07]::Super) }, ItemId { owner_id: DefId(0:8 ~ a[ad07]::Dyn) }, ItemId { owner_id: DefId(0:9 ~ a[ad07]::{impl#0}) }, ItemId { owner_id: DefId(0:12 ~ a[ad07]::Foo) }, ItemId { owner_id: DefId(0:16 ~ a[ad07]::Mirror) }, ItemId { owner_id: DefId(0:18 ~ a[ad07]::{impl#1}) }] })', 'error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:212:14: generics_of: unexpected node kind Crate(Mod { spans: ModSpans { inner_span: /tmp/im/a.rs:1:1: 20:57 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ a[ad07]::std) }, ItemId { owner_id: DefId(0:2 ~ a[ad07]::{use#0}) }, ItemId { owner_id: DefId(0:3 ~ a[ad07]::ActuallySuper) }, ItemId { owner_id: DefId(0:5 ~ a[ad07]::Super) }, ItemId { owner_id: DefId(0:8 ~ a[ad07]::Dyn) }, ItemId { owner_id: DefId(0:9 ~ a[ad07]::{impl#0}) }, ItemId { owner_id: DefId(0:12 ~ a[ad07]::Foo) }, ItemId { owner_id: DefId(0:16 ~ a[ad07]::Mirror) }, ItemId { owner_id: DefId(0:18 ~ a[ad07]::{impl#1}) }] })'
File: /tmp/im/a.rs
-->

snippet:
````rust
struct ActuallySuper;

trait Super<B = Self> {
    type Assoc;
}

trait Dyn {}
impl<T, U> Dyn for dyn Foo<T, U> + '_ {}

trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
where
    <Self as Mirror>::Assoc: Super,
{
    fn transmute(&self, t: T) -> <Self as B>::Assoc;
}

trait Mirror {
    type Assoc: ?Sized;
}
impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}

````

Version information
````
rustc 1.95.0-nightly (13c38730d 2026-02-08)
binary: rustc
commit-hash: 13c38730d981289cc7ae4cc109fd7756bf83ee67
commit-date: 2026-02-08
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 22.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/13c38730d981289cc7ae4cc109fd7756bf83ee67/compiler/rustc_hir_analysis/src/collect/generics_of.rs#L206-L218

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0405]: cannot find trait `B` in this scope
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:14:43
   |
14 |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
   |                                           ^ not found in this scope

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:57
   |
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   |                                                         ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs`

error: internal compiler error: /rustc-dev/13c38730d981289cc7ae4cc109fd7756bf83ee67/compiler/rustc_hir_analysis/src/collect/generics_of.rs:212:14: generics_of: unexpected node kind Crate(Mod { spans: ModSpans { inner_span: /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:1:1: 20:57 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ mvce[ba04]::std) }, ItemId { owner_id: DefId(0:2 ~ mvce[ba04]::{use#0}) }, ItemId { owner_id: DefId(0:3 ~ mvce[ba04]::ActuallySuper) }, ItemId { owner_id: DefId(0:5 ~ mvce[ba04]::Super) }, ItemId { owner_id: DefId(0:8 ~ mvce[ba04]::Dyn) }, ItemId { owner_id: DefId(0:9 ~ mvce[ba04]::{impl#0}) }, ItemId { owner_id: DefId(0:12 ~ mvce[ba04]::Foo) }, ItemId { owner_id: DefId(0:16 ~ mvce[ba04]::Mirror) }, ItemId { owner_id: DefId(0:18 ~ mvce[ba04]::{impl#1}) }] })
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:1:1
   |
 1 | / struct ActuallySuper;
 2 | |
 3 | | trait Super<B = Self> {
 4 | |     type Assoc;
...  |
20 | | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   | |________________________________________________________^


thread 'rustc' (3909009) panicked at /rustc-dev/13c38730d981289cc7ae4cc109fd7756bf83ee67/compiler/rustc_hir_analysis/src/collect/generics_of.rs:212:14:
Box<dyn Any>
stack backtrace:
   0:     0x7f655e7bebcb - <<std[bd410f491febadb6]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[2ca3c49344714521]::fmt::Display>::fmt
   1:     0x7f655ee0ecc8 - core[2ca3c49344714521]::fmt::write
   2:     0x7f655e7d5eb6 - <std[bd410f491febadb6]::sys::stdio::unix::Stderr as std[bd410f491febadb6]::io::Write>::write_fmt
   3:     0x7f655e794c08 - std[bd410f491febadb6]::panicking::default_hook::{closure#0}
   4:     0x7f655e7b1fd3 - std[bd410f491febadb6]::panicking::default_hook
   5:     0x7f655d78c9e2 - std[bd410f491febadb6]::panicking::update_hook::<alloc[a248f8e1b1583859]::boxed::Box<rustc_driver_impl[b59eeeb9fc736c31]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f655e7b22b2 - std[bd410f491febadb6]::panicking::panic_with_hook
   7:     0x7f655d7c30f1 - std[bd410f491febadb6]::panicking::begin_panic::<rustc_errors[9a7d3fb2b4d26c7d]::ExplicitBug>::{closure#0}
   8:     0x7f655d7ba246 - std[bd410f491febadb6]::sys::backtrace::__rust_end_short_backtrace::<std[bd410f491febadb6]::panicking::begin_panic<rustc_errors[9a7d3fb2b4d26c7d]::ExplicitBug>::{closure#0}, !>
   9:     0x7f655d7b9fab - std[bd410f491febadb6]::panicking::begin_panic::<rustc_errors[9a7d3fb2b4d26c7d]::ExplicitBug>
  10:     0x7f655d7cee61 - <rustc_errors[9a7d3fb2b4d26c7d]::diagnostic::BugAbort as rustc_errors[9a7d3fb2b4d26c7d]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7f655dd5d4ac - <rustc_errors[9a7d3fb2b4d26c7d]::DiagCtxtHandle>::span_bug::<rustc_span[6275488d32054521]::span_encoding::Span, alloc[a248f8e1b1583859]::string::String>
  12:     0x7f655dd85196 - rustc_middle[3efcf2d0f50fae3f]::util::bug::opt_span_bug_fmt::<rustc_span[6275488d32054521]::span_encoding::Span>::{closure#0}
  13:     0x7f655dd85342 - rustc_middle[3efcf2d0f50fae3f]::ty::context::tls::with_opt::<rustc_middle[3efcf2d0f50fae3f]::util::bug::opt_span_bug_fmt<rustc_span[6275488d32054521]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:     0x7f655dd7694b - rustc_middle[3efcf2d0f50fae3f]::ty::context::tls::with_context_opt::<rustc_middle[3efcf2d0f50fae3f]::ty::context::tls::with_opt<rustc_middle[3efcf2d0f50fae3f]::util::bug::opt_span_bug_fmt<rustc_span[6275488d32054521]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:     0x7f655ba636f8 - rustc_middle[3efcf2d0f50fae3f]::util::bug::span_bug_fmt::<rustc_span[6275488d32054521]::span_encoding::Span>
  16:     0x7f655f14a5fb - rustc_hir_analysis[9091eb4b622a99a6]::collect::generics_of::generics_of
  17:     0x7f655f148162 - rustc_query_impl[21826ead98e520a1]::query_impl::generics_of::compute_fn::__rust_begin_short_backtrace
  18:     0x7f655f15569e - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::DefIdCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 8usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: true }}>, false>
  19:     0x7f655f1553dd - rustc_query_impl[21826ead98e520a1]::query_impl::generics_of::get_query_non_incr::__rust_end_short_backtrace
  20:     0x7f655e579526 - <rustc_trait_selection[f7d673c3e0fcaf59]::error_reporting::TypeErrCtxt>::note_and_explain_type_err
  21:     0x7f655d060f1b - <rustc_trait_selection[f7d673c3e0fcaf59]::error_reporting::TypeErrCtxt>::note_type_err
  22:     0x7f655d098687 - <rustc_trait_selection[f7d673c3e0fcaf59]::error_reporting::TypeErrCtxt>::report_fulfillment_errors
  23:     0x7f655f783c97 - rustc_trait_selection[f7d673c3e0fcaf59]::traits::do_normalize_predicates
  24:     0x7f655f7e75bf - rustc_trait_selection[f7d673c3e0fcaf59]::traits::normalize_param_env_or_error
  25:     0x7f655fe136c6 - rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::virtual_call_violations_for_method
  26:     0x7f655fe18eec - rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::dyn_compatibility_violations_for_assoc_item
  27:     0x7f655fe18b98 - <core[2ca3c49344714521]::iter::adapters::flatten::FlatMap<core[2ca3c49344714521]::iter::adapters::map::Map<core[2ca3c49344714521]::iter::adapters::map::Map<core[2ca3c49344714521]::slice::iter::Iter<(core[2ca3c49344714521]::option::Option<rustc_span[6275488d32054521]::symbol::Symbol>, rustc_middle[3efcf2d0f50fae3f]::ty::assoc::AssocItem)>, <rustc_data_structures[ca9174053067dbfd]::sorted_map::index_map::SortedIndexMultiMap<u32, core[2ca3c49344714521]::option::Option<rustc_span[6275488d32054521]::symbol::Symbol>, rustc_middle[3efcf2d0f50fae3f]::ty::assoc::AssocItem>>::iter::{closure#0}>, <rustc_middle[3efcf2d0f50fae3f]::ty::assoc::AssocItems>::in_definition_order::{closure#0}>, alloc[a248f8e1b1583859]::vec::Vec<rustc_middle[3efcf2d0f50fae3f]::traits::DynCompatibilityViolation>, rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::dyn_compatibility_violations_for_trait::{closure#0}::{closure#0}> as core[2ca3c49344714521]::iter::traits::iterator::Iterator>::next
  28:     0x7f655fe16b6e - <core[2ca3c49344714521]::iter::adapters::GenericShunt<core[2ca3c49344714521]::iter::adapters::map::Map<core[2ca3c49344714521]::iter::adapters::flatten::FlatMap<core[2ca3c49344714521]::iter::sources::from_fn::FromFn<rustc_type_ir[b71a8d5e563732d6]::elaborate::supertrait_def_ids<rustc_middle[3efcf2d0f50fae3f]::ty::context::TyCtxt>::{closure#0}>, alloc[a248f8e1b1583859]::vec::Vec<rustc_middle[3efcf2d0f50fae3f]::traits::DynCompatibilityViolation>, rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::dyn_compatibility_violations::{closure#0}>, core[2ca3c49344714521]::result::Result<rustc_middle[3efcf2d0f50fae3f]::traits::DynCompatibilityViolation, !>::Ok>, core[2ca3c49344714521]::result::Result<core[2ca3c49344714521]::convert::Infallible, !>> as core[2ca3c49344714521]::iter::traits::iterator::Iterator>::next
  29:     0x7f655fe15f3d - rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::dyn_compatibility_violations
  30:     0x7f655fe15e14 - rustc_query_impl[21826ead98e520a1]::query_impl::dyn_compatibility_violations::compute_fn::__rust_begin_short_backtrace
  31:     0x7f655f129f27 - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::DefIdCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 16usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  32:     0x7f655f88ec1e - rustc_query_impl[21826ead98e520a1]::query_impl::dyn_compatibility_violations::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f655f88ecbc - rustc_trait_selection[f7d673c3e0fcaf59]::traits::dyn_compatibility::is_dyn_compatible
  34:     0x7f655f88ec53 - rustc_query_impl[21826ead98e520a1]::query_impl::is_dyn_compatible::compute_fn::__rust_begin_short_backtrace
  35:     0x7f655f87e892 - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::DefIdCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 1usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  36:     0x7f655f87d4c6 - rustc_query_impl[21826ead98e520a1]::query_impl::is_dyn_compatible::get_query_non_incr::__rust_end_short_backtrace
  37:     0x7f655ef9ded7 - rustc_hir_analysis[9091eb4b622a99a6]::coherence::coherent_trait
  38:     0x7f655ef9d123 - rustc_query_impl[21826ead98e520a1]::query_impl::coherent_trait::compute_fn::__rust_begin_short_backtrace
  39:     0x7f655f87e892 - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::DefIdCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 1usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  40:     0x7f655f87df24 - rustc_query_impl[21826ead98e520a1]::query_impl::coherent_trait::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7f655f4dd334 - rustc_hir_analysis[9091eb4b622a99a6]::check::check::check_item_type
  42:     0x7f655f4db63a - rustc_hir_analysis[9091eb4b622a99a6]::check::wfcheck::check_well_formed
  43:     0x7f655f4db61b - rustc_query_impl[21826ead98e520a1]::query_impl::check_well_formed::compute_fn::__rust_begin_short_backtrace
  44:     0x7f655f4db03a - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_data_structures[ca9174053067dbfd]::vec_cache::VecCache<rustc_span[6275488d32054521]::def_id::LocalDefId, rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 1usize]>, rustc_query_system[6518298d71496d73]::dep_graph::graph::DepNodeIndex>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  45:     0x7f655f4dab82 - rustc_query_impl[21826ead98e520a1]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  46:     0x7f655f4d8269 - rustc_hir_analysis[9091eb4b622a99a6]::check::wfcheck::check_type_wf
  47:     0x7f655f4d8163 - rustc_query_impl[21826ead98e520a1]::query_impl::check_type_wf::compute_fn::__rust_begin_short_backtrace
  48:     0x7f6560093487 - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::SingleCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 1usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  49:     0x7f6560093136 - rustc_query_impl[21826ead98e520a1]::query_impl::check_type_wf::get_query_non_incr::__rust_end_short_backtrace
  50:     0x7f655fac6e73 - rustc_hir_analysis[9091eb4b622a99a6]::check_crate
  51:     0x7f655fabf359 - rustc_interface[923b877a2cdf5a67]::passes::analysis
  52:     0x7f65600903ce - rustc_query_system[6518298d71496d73]::query::plumbing::try_execute_query::<rustc_query_impl[21826ead98e520a1]::SemiDynamicQueryDispatcher<rustc_query_system[6518298d71496d73]::query::caches::SingleCache<rustc_middle[3efcf2d0f50fae3f]::query::erase::ErasedData<[u8; 0usize]>>, {rustc_query_impl[21826ead98e520a1]::QueryFlags { is_anon: false, is_depth_limit: false, is_feedable: false }}>, false>
  53:     0x7f6560090192 - rustc_query_impl[21826ead98e520a1]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  54:     0x7f65601d32e7 - <rustc_interface[923b877a2cdf5a67]::passes::create_and_enter_global_ctxt<core[2ca3c49344714521]::option::Option<rustc_interface[923b877a2cdf5a67]::queries::Linker>, rustc_driver_impl[b59eeeb9fc736c31]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[2ca3c49344714521]::ops::function::FnOnce<(&rustc_session[45e3122b25e742c0]::session::Session, rustc_middle[3efcf2d0f50fae3f]::ty::context::CurrentGcx, alloc[a248f8e1b1583859]::sync::Arc<rustc_data_structures[ca9174053067dbfd]::jobserver::Proxy>, &std[bd410f491febadb6]::sync::once_lock::OnceLock<rustc_middle[3efcf2d0f50fae3f]::ty::context::GlobalCtxt>, &rustc_data_structures[ca9174053067dbfd]::sync::worker_local::WorkerLocal<rustc_middle[3efcf2d0f50fae3f]::arena::Arena>, &rustc_data_structures[ca9174053067dbfd]::sync::worker_local::WorkerLocal<rustc_hir[6cfa7645bd0a50ff]::Arena>, rustc_driver_impl[b59eeeb9fc736c31]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  55:     0x7f656007e261 - rustc_interface[923b877a2cdf5a67]::interface::run_compiler::<(), rustc_driver_impl[b59eeeb9fc736c31]::run_compiler::{closure#0}>::{closure#1}
  56:     0x7f65600601fe - std[bd410f491febadb6]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[923b877a2cdf5a67]::util::run_in_thread_with_globals<rustc_interface[923b877a2cdf5a67]::util::run_in_thread_pool_with_globals<rustc_interface[923b877a2cdf5a67]::interface::run_compiler<(), rustc_driver_impl[b59eeeb9fc736c31]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  57:     0x7f6560060720 - <std[bd410f491febadb6]::thread::lifecycle::spawn_unchecked<rustc_interface[923b877a2cdf5a67]::util::run_in_thread_with_globals<rustc_interface[923b877a2cdf5a67]::util::run_in_thread_pool_with_globals<rustc_interface[923b877a2cdf5a67]::interface::run_compiler<(), rustc_driver_impl[b59eeeb9fc736c31]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[2ca3c49344714521]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  58:     0x7f65600615ac - <std[bd410f491febadb6]::sys::thread::unix::Thread>::new::thread_start
  59:     0x7f655989698b - <unknown>
  60:     0x7f655991a9cc - <unknown>
  61:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.95.0-nightly (13c38730d 2026-02-08) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [generics_of] computing generics of ``
#1 [dyn_compatibility_violations] determining dyn-compatibility of trait `Foo`
#2 [is_dyn_compatible] checking if trait `Foo` is dyn-compatible
#3 [coherent_trait] coherence checking all impls of trait `Dyn`
#4 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:8:1: 8:38>` is well-formed
#5 [check_type_wf] checking that types are well-formed
#6 [analysis] running analysis passes on crate `mvce`
end of query stack
error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:10:1
   |
10 | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `Self`, found type parameter `T`
   |
   = note: expected type parameter `Self`
              found type parameter `T`
   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
note: required for `Self` to implement `Mirror`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:42
   |
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   |                              ---------   ^^^^^^     ^
   |                              |
   |                              unsatisfied trait bound introduced here

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:10:1
   |
10 | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
note: required for `Self` to implement `Mirror`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:42
   |
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   |      -                                   ^^^^^^     ^
   |      |
   |      unsatisfied trait bound implicitly introduced here
help: consider further restricting `Self`
   |
10 | trait Foo<T, U>: Super<ActuallySuper, Assoc = T> + Sized
   |                                                  +++++++

error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:12:30
   |
10 | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
   | ------------------------------------------------
   | |         |
   | |         found type parameter
   | expected type parameter
11 | where
12 |     <Self as Mirror>::Assoc: Super,
   |                              ^^^^^ expected type parameter `Self`, found type parameter `T`
   |
   = note: expected type parameter `Self`
              found type parameter `T`
   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
note: required for `Self` to implement `Mirror`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:42
   |
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   |                              ---------   ^^^^^^     ^
   |                              |
   |                              unsatisfied trait bound introduced here

error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:12:30
   |
12 |     <Self as Mirror>::Assoc: Super,
   |                              ^^^^^ doesn't have a size known at compile-time
   |
note: required for `Self` to implement `Mirror`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:42
   |
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   |      -                                   ^^^^^^     ^
   |      |
   |      unsatisfied trait bound implicitly introduced here
help: consider further restricting `Self`
   |
10 | trait Foo<T, U>: Super<ActuallySuper, Assoc = T> + Sized
   |                                                  +++++++

error[E0046]: not all trait items implemented, missing: `Assoc`
  --> /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:20:1
   |
18 |     type Assoc: ?Sized;
   |     ------------------ `Assoc` from trait
19 | }
20 | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Assoc` in implementation

error: aborting due to 8 previous errors

Some errors have detailed explanations: E0046, E0271, E0277, E0405, E0601.
For more information about an error, try `rustc --explain E0046`.

```

</p>
</details>

<!--
query stack:
error: internal compiler error: /rustc-dev/13c38730d981289cc7ae4cc109fd7756bf83ee67/compiler/rustc_hir_analysis/src/collect/generics_of.rs:212:14: generics_of: unexpected node kind Crate(Mod { spans: ModSpans { inner_span: /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:1:1: 20:57 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ mvce[ba04]::std) }, ItemId { owner_id: DefId(0:2 ~ mvce[ba04]::{use#0}) }, ItemId { owner_id: DefId(0:3 ~ mvce[ba04]::ActuallySuper) }, ItemId { owner_id: DefId(0:5 ~ mvce[ba04]::Super) }, ItemId { owner_id: DefId(0:8 ~ mvce[ba04]::Dyn) }, ItemId { owner_id: DefId(0:9 ~ mvce[ba04]::{impl#0}) }, ItemId { owner_id: DefId(0:12 ~ mvce[ba04]::Foo) }, ItemId { owner_id: DefId(0:16 ~ mvce[ba04]::Mirror) }, ItemId { owner_id: DefId(0:18 ~ mvce[ba04]::{impl#1}) }] })
#0 [generics_of] computing generics of ``
#1 [dyn_compatibility_violations] determining dyn-compatibility of trait `Foo`
#2 [is_dyn_compatible] checking if trait `Foo` is dyn-compatible
#3 [coherent_trait] coherence checking all impls of trait `Dyn`
#4 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.8QRXymLaEPJH/rustc_testrunner_tmpdir_reporting.jo3x9uMFrVfW/mvce.rs:8:1: 8:38>` is well-formed
#5 [check_type_wf] checking that types are well-formed
#6 [analysis] running analysis passes on crate `mvce`

-->



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs
@@ -55,7 +55,7 @@ use rustc_data_structures::fx::{FxIndexMap, FxIndexSet};
 use rustc_errors::{Applicability, Diag, DiagStyledString, IntoDiagArg, StringPart, pluralize};
 use rustc_hir as hir;
 use rustc_hir::def::DefKind;
-use rustc_hir::def_id::DefId;
+use rustc_hir::def_id::{CRATE_DEF_ID, DefId};
 use rustc_hir::intravisit::Visitor;
 use rustc_hir::lang_items::LangItem;
 use rustc_infer::infer::DefineOpaqueTypes;
@@ -1793,12 +1793,13 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }
         }
 
-        self.note_and_explain_type_err(diag, terr, cause, span, cause.body_id.to_def_id());
+        let body_owner_def_id = (cause.body_id != CRATE_DEF_ID).then(|| cause.body_id.to_def_id());
+        self.note_and_explain_type_err(diag, terr, cause, span, body_owner_def_id);
         if let Some(exp_found) = exp_found
             && let exp_found = TypeError::Sorts(exp_found)
             && exp_found != terr
         {
-            self.note_and_explain_type_err(diag, exp_found, cause, span, cause.body_id.to_def_id());
+            self.note_and_explain_type_err(diag, exp_found, cause, span, body_owner_def_id);
         }
 
         if let Some(ValuePairs::TraitRefs(exp_found)) = values

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
@@ -22,11 +22,13 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         err: TypeError<'tcx>,
         cause: &ObligationCause<'tcx>,
         sp: Span,
-        body_owner_def_id: DefId,
+        body_owner_def_id: Option<DefId>,
     ) {
         debug!("note_and_explain_type_err err={:?} cause={:?}", err, cause);
 
         let tcx = self.tcx;
+
+        let body_generics = body_owner_def_id.map(|def_id| tcx.generics_of(def_id));
 
         match err {
             TypeError::ArgumentSorts(values, _) | TypeError::Sorts(values) => {
@@ -64,14 +66,15 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         }
                     }
                     (ty::Param(expected), ty::Param(found)) => {
-                        let generics = tcx.generics_of(body_owner_def_id);
-                        let e_span = tcx.def_span(generics.type_param(expected, tcx).def_id);
-                        if !sp.contains(e_span) {
-                            diag.span_label(e_span, "expected type parameter");
-                        }
-                        let f_span = tcx.def_span(generics.type_param(found, tcx).def_id);
-                        if !sp.contains(f_span) {
-                            diag.span_label(f_span, "found type parameter");
+                        if let Some(generics) = body_generics {
+                            let e_span = tcx.def_span(generics.type_param(expected, tcx).def_id);
+                            if !sp.contains(e_span) {
+                                diag.span_label(e_span, "expected type parameter");
+                            }
+                            let f_span = tcx.def_span(generics.type_param(found, tcx).def_id);
+                            if !sp.contains(f_span) {
+                                diag.span_label(f_span, "found type parameter");
+                            }
                         }
                         diag.note(
                             "a type parameter was expected, but a different one was found; \
@@ -92,9 +95,10 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     // FIXME(inherent_associated_types): Extend this to support `ty::Inherent`, too.
                     (ty::Param(p), ty::Alias(ty::Projection, proj))
                     | (ty::Alias(ty::Projection, proj), ty::Param(p))
-                        if !tcx.is_impl_trait_in_trait(proj.def_id) =>
+                        if !tcx.is_impl_trait_in_trait(proj.def_id)
+                            && let Some(generics) = body_generics =>
                     {
-                        let param = tcx.generics_of(body_owner_def_id).type_param(p, tcx);
+                        let param = generics.type_param(p, tcx);
                         let p_def_id = param.def_id;
                         let p_span = tcx.def_span(p_def_id);
                         let expected = match (values.expected.kind(), values.found.kind()) {
@@ -106,9 +110,7 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             diag.span_label(p_span, format!("{expected}this type parameter"));
                         }
                         let param_def_id = match *proj.self_ty().kind() {
-                            ty::Param(param) => {
-                                tcx.generics_of(body_owner_def_id).type_param(param, tcx).def_id
-                            }
+                            ty::Param(param) => generics.type_param(param, tcx).def_id,
                             _ => p_def_id,
                         };
                         let parent = param_def_id.as_local().and_then(|id| {
@@ -188,15 +190,16 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     }
                     (ty::Param(p), ty::Dynamic(..) | ty::Alias(ty::Opaque, ..))
                     | (ty::Dynamic(..) | ty::Alias(ty::Opaque, ..), ty::Param(p)) => {
-                        let generics = tcx.generics_of(body_owner_def_id);
-                        let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
-                        let expected = match (values.expected.kind(), values.found.kind()) {
-                            (ty::Param(_), _) => "expected ",
-                            (_, ty::Param(_)) => "found ",
-                            _ => "",
-                        };
-                        if !sp.contains(p_span) {
-                            diag.span_label(p_span, format!("{expected}this type parameter"));
+                        if let Some(generics) = body_generics {
+                            let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
+                            let expected = match (values.expected.kind(), values.found.kind()) {
+                                (ty::Param(_), _) => "expected ",
+                                (_, ty::Param(_)) => "found ",
+                                _ => "",
+                            };
+                            if !sp.contains(p_span) {
+                                diag.span_label(p_span, format!("{expected}this type parameter"));
+                            }
                         }
                         diag.help("type parameters must be constrained to match other types");
                         if diag.code.is_some_and(|code| tcx.sess.teach(code)) {
@@ -236,18 +239,18 @@ impl<T> Trait<T> for X {
                         ty::Param(p),
                         ty::Closure(..) | ty::CoroutineClosure(..) | ty::Coroutine(..),
                     ) => {
-                        let generics = tcx.generics_of(body_owner_def_id);
-                        let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
-                        if !sp.contains(p_span) {
-                            diag.span_label(p_span, "expected this type parameter");
+                        if let Some(generics) = body_generics {
+                            let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
+                            if !sp.contains(p_span) {
+                                diag.span_label(p_span, "expected this type parameter");
+                            }
                         }
                         diag.help(format!(
                             "every closure has a distinct type and so could not always match the \
                              caller-chosen type of parameter `{p}`"
                         ));
                     }
-                    (ty::Param(p), _) | (_, ty::Param(p)) => {
-                        let generics = tcx.generics_of(body_owner_def_id);
+                    (ty::Param(p), _) | (_, ty::Param(p)) if let Some(generics) = body_generics => {
                         let p_span = tcx.def_span(generics.type_param(p, tcx).def_id);
                         let expected = match (values.expected.kind(), values.found.kind()) {
                             (ty::Param(_), _) => "expected ",
@@ -366,7 +369,8 @@ impl<T> Trait<T> for X {
                     }
                     (_, ty::Alias(ty::Opaque, opaque_ty))
                     | (ty::Alias(ty::Opaque, opaque_ty), _) => {
-                        if opaque_ty.def_id.is_local()
+                        if let Some(body_owner_def_id) = body_owner_def_id
+                            && opaque_ty.def_id.is_local()
                             && matches!(
                                 tcx.def_kind(body_owner_def_id),
                                 DefKind::Fn
@@ -560,11 +564,14 @@ impl<T> Trait<T> for X {
         &self,
         diag: &mut Diag<'_>,
         msg: impl Fn() -> String,
-        body_owner_def_id: DefId,
+        body_owner_def_id: Option<DefId>,
         proj_ty: ty::AliasTy<'tcx>,
         ty: Ty<'tcx>,
     ) -> bool {
         let tcx = self.tcx;
+        let Some(body_owner_def_id) = body_owner_def_id else {
+            return false;
+        };
         let assoc = tcx.associated_item(proj_ty.def_id);
         let (trait_ref, assoc_args) = proj_ty.trait_ref_and_own_args(tcx);
         let Some(item) = tcx.hir_get_if_local(body_owner_def_id) else {
@@ -611,7 +618,7 @@ impl<T> Trait<T> for X {
             _ => return false,
         };
         let parent = tcx.hir_get_parent_item(hir_id).def_id;
-        self.suggest_constraint(diag, msg, parent.into(), proj_ty, ty)
+        self.suggest_constraint(diag, msg, Some(parent.into()), proj_ty, ty)
     }
 
     /// An associated type was expected and a different type was found.
@@ -632,7 +639,7 @@ impl<T> Trait<T> for X {
         diag: &mut Diag<'_>,
         proj_ty: ty::AliasTy<'tcx>,
         values: ExpectedFound<Ty<'tcx>>,
-        body_owner_def_id: DefId,
+        body_owner_def_id: Option<DefId>,
         cause_code: &ObligationCauseCode<'_>,
     ) {
         let tcx = self.tcx;
@@ -653,7 +660,7 @@ impl<T> Trait<T> for X {
             )
         };
 
-        let body_owner = tcx.hir_get_if_local(body_owner_def_id);
+        let body_owner = body_owner_def_id.and_then(|id| tcx.hir_get_if_local(id));
         let current_method_ident = body_owner.and_then(|n| n.ident()).map(|i| i.name);
 
         // We don't want to suggest calling an assoc fn in a scope where that isn't feasible.
@@ -835,12 +842,12 @@ fn foo(&self) -> Self::T { String::new() }
     fn point_at_associated_type(
         &self,
         diag: &mut Diag<'_>,
-        body_owner_def_id: DefId,
+        body_owner_def_id: Option<DefId>,
         found: Ty<'tcx>,
     ) -> bool {
         let tcx = self.tcx;
 
-        let Some(def_id) = body_owner_def_id.as_local() else {
+        let Some(def_id) = body_owner_def_id.and_then(|id| id.as_local()) else {
             return false;
         };
 
@@ -852,7 +859,7 @@ fn foo(&self) -> Self::T { String::new() }
 
         debug!("expected_projection parent item {:?}", item);
 
-        let param_env = tcx.param_env(body_owner_def_id);
+        let param_env = tcx.param_env(def_id);
 
         if let DefKind::Trait | DefKind::Impl { .. } = tcx.def_kind(parent_id) {
             let assoc_items = tcx.associated_items(parent_id);

--- a/tests/ui/dyn-compatibility/ice-generics-of-crate-root-152335.rs
+++ b/tests/ui/dyn-compatibility/ice-generics-of-crate-root-152335.rs
@@ -1,0 +1,40 @@
+// Regression test for #152335.
+// The compiler used to ICE in `generics_of` when `note_and_explain_type_err`
+// was called with `CRATE_DEF_ID` as the body owner during dyn-compatibility
+// checking. This happened because `ObligationCause::dummy()` sets `body_id`
+// to `CRATE_DEF_ID`, and error reporting tried to look up generics on it.
+
+struct ActuallySuper;
+
+trait Super<B = Self> {
+    type Assoc;
+}
+
+trait Dyn {}
+impl<T, U> Dyn for dyn Foo<T, U> + '_ {}
+//~^ ERROR the trait `Foo` is not dyn compatible
+
+trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+//~^ ERROR type mismatch resolving
+//~| ERROR the size for values of type `Self` cannot be known
+where
+    <Self as Mirror>::Assoc: Super,
+    //~^ ERROR type mismatch resolving
+    //~| ERROR the size for values of type `Self` cannot be known
+    //~| ERROR type mismatch resolving
+    //~| ERROR the size for values of type `Self` cannot be known
+{
+    fn transmute(&self, t: T) -> <Self as B>::Assoc;
+    //~^ ERROR cannot find trait `B` in this scope
+    //~| ERROR type mismatch resolving
+    //~| ERROR the size for values of type `Self` cannot be known
+    //~| ERROR type mismatch resolving
+    //~| ERROR the size for values of type `Self` cannot be known
+}
+
+trait Mirror {
+    type Assoc: ?Sized;
+}
+impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+//~^ ERROR not all trait items implemented
+//~| ERROR `main` function not found

--- a/tests/ui/dyn-compatibility/ice-generics-of-crate-root-152335.stderr
+++ b/tests/ui/dyn-compatibility/ice-generics-of-crate-root-152335.stderr
@@ -1,0 +1,236 @@
+error[E0405]: cannot find trait `B` in this scope
+  --> $DIR/ice-generics-of-crate-root-152335.rs:27:43
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
+   |                                           ^ not found in this scope
+
+error[E0601]: `main` function not found in crate `ice_generics_of_crate_root_152335`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:57
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                                                         ^ consider adding a `main` function to `$DIR/ice-generics-of-crate-root-152335.rs`
+
+error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:27:5
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `Self`, found type parameter `T`
+   |
+   = note: expected type parameter `Self`
+              found type parameter `T`
+   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
+   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                              ---------   ^^^^^^     ^
+   |                              |
+   |                              unsatisfied trait bound introduced here
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/ice-generics-of-crate-root-152335.rs:27:5
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |      -                                   ^^^^^^     ^
+   |      |
+   |      unsatisfied trait bound implicitly introduced here
+
+error[E0038]: the trait `Foo` is not dyn compatible
+  --> $DIR/ice-generics-of-crate-root-152335.rs:14:20
+   |
+LL | impl<T, U> Dyn for dyn Foo<T, U> + '_ {}
+   |                    ^^^^^^^^^^^^^^^^^^ `Foo` is not dyn compatible
+   |
+note: for a trait to be dyn compatible it needs to allow building a vtable
+      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
+  --> $DIR/ice-generics-of-crate-root-152335.rs:21:30
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   |       --- this trait is not dyn compatible...
+...
+LL |     <Self as Mirror>::Assoc: Super,
+   |                              ^^^^^ ...because it uses `Self` as a type parameter
+
+error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:17:1
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `Self`, found type parameter `T`
+   |
+   = note: expected type parameter `Self`
+              found type parameter `T`
+   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
+   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                              ---------   ^^^^^^     ^
+   |                              |
+   |                              unsatisfied trait bound introduced here
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/ice-generics-of-crate-root-152335.rs:17:1
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |      -                                   ^^^^^^     ^
+   |      |
+   |      unsatisfied trait bound implicitly introduced here
+help: consider further restricting `Self`
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T> + Sized
+   |                                                  +++++++
+
+error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:21:30
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   | ------------------------------------------------
+   | |         |
+   | |         found type parameter
+   | expected type parameter
+...
+LL |     <Self as Mirror>::Assoc: Super,
+   |                              ^^^^^ expected type parameter `Self`, found type parameter `T`
+   |
+   = note: expected type parameter `Self`
+              found type parameter `T`
+   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
+   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                              ---------   ^^^^^^     ^
+   |                              |
+   |                              unsatisfied trait bound introduced here
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/ice-generics-of-crate-root-152335.rs:21:30
+   |
+LL |     <Self as Mirror>::Assoc: Super,
+   |                              ^^^^^ doesn't have a size known at compile-time
+   |
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |      -                                   ^^^^^^     ^
+   |      |
+   |      unsatisfied trait bound implicitly introduced here
+help: consider further restricting `Self`
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T> + Sized
+   |                                                  +++++++
+
+error[E0046]: not all trait items implemented, missing: `Assoc`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:1
+   |
+LL |     type Assoc: ?Sized;
+   |     ------------------ `Assoc` from trait
+LL | }
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Assoc` in implementation
+
+error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:27:5
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   | ------------------------------------------------
+   | |         |
+   | |         found type parameter
+   | expected type parameter
+...
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `Self`, found type parameter `T`
+   |
+   = note: expected type parameter `Self`
+              found type parameter `T`
+   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
+   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                              ---------   ^^^^^^     ^
+   |                              |
+   |                              unsatisfied trait bound introduced here
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/ice-generics-of-crate-root-152335.rs:27:5
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |      -                                   ^^^^^^     ^
+   |      |
+   |      unsatisfied trait bound implicitly introduced here
+help: consider further restricting `Self`
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc where Self: Sized;
+   |                                                     +++++++++++++++++
+
+error[E0271]: type mismatch resolving `<Self as Super<ActuallySuper>>::Assoc == Self`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:21:30
+   |
+LL | trait Foo<T, U>: Super<ActuallySuper, Assoc = T>
+   | ------------------------------------------------
+   | |         |
+   | |         found type parameter
+   | expected type parameter
+...
+LL |     <Self as Mirror>::Assoc: Super,
+   |                              ^^^^^ expected type parameter `Self`, found type parameter `T`
+   |
+   = note: expected type parameter `Self`
+              found type parameter `T`
+   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
+   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |                              ---------   ^^^^^^     ^
+   |                              |
+   |                              unsatisfied trait bound introduced here
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/ice-generics-of-crate-root-152335.rs:21:30
+   |
+LL |     <Self as Mirror>::Assoc: Super,
+   |                              ^^^^^ doesn't have a size known at compile-time
+   |
+note: required for `Self` to implement `Mirror`
+  --> $DIR/ice-generics-of-crate-root-152335.rs:38:42
+   |
+LL | impl<T: Super<ActuallySuper, Assoc = T>> Mirror for T {}
+   |      -                                   ^^^^^^     ^
+   |      |
+   |      unsatisfied trait bound implicitly introduced here
+help: consider further restricting `Self`
+   |
+LL |     fn transmute(&self, t: T) -> <Self as B>::Assoc where Self: Sized;
+   |                                                     +++++++++++++++++
+
+error: aborting due to 14 previous errors
+
+Some errors have detailed explanations: E0038, E0046, E0271, E0277, E0405, E0601.
+For more information about an error, try `rustc --explain E0038`.

